File: backend/app/schemas/research_job.py
"""
app/schemas/research_job.py — Pydantic request / response schemas for research jobs.
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, field_validator


class JobCreate(BaseModel):
    """Payload used to create a new research job."""

    topic: str = Field(
        ...,
        min_length=3,
        max_length=512,
        description="The research topic to investigate.",
        examples=["quantum computing", "climate change mitigation strategies"],
    )
    depth_minutes: int = Field(
        default=5,
        ge=1,
        le=60,
        description="Intended research depth in minutes (1–60). "
                    "Reserved for future pipeline time-bounding logic.",
    )

    @field_validator("topic")
    @classmethod
    def topic_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("topic must not be blank")
        return v.strip()


class JobResponse(BaseModel):
    """Response schema returned for research job reads."""

    id: int
    topic: str
    depth_minutes: int
    status: str = Field(
        description=(
            "Pipeline stage. One of: queued, retrieving_sources, "
            "fetching_documents, chunking_documents, embedding_documents, "
            "drafting_outline, writing_report, completed, failed."
        )
    )
    progress: int = Field(description="Completion percentage (0–100).")
    result_report: Optional[str] = Field(
        default=None,
        description="The generated research report. Non-null only when status=completed.",
    )
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}

File: Claude_code/backend_connected/files/research_job.dart
/// Mirrors the backend `JobResponse` Pydantic schema.
///
/// Backend fields (snake_case) → Dart fields (camelCase).
class ResearchJob {
  final int id;
  final String topic;

  /// Corresponds to `depth_minutes` in the backend.
  final int depthMinutes;

  /// Pipeline stage: queued | retrieving_sources | fetching_documents |
  /// chunking_documents | embedding_documents | drafting_outline |
  /// writing_report | completed | failed
  final String status;

  /// Completion percentage 0–100.
  final int progress;

  /// The generated report text — non-null only when status == 'completed'.
  final String? resultReport;

  final DateTime createdAt;
  final DateTime? updatedAt;

  const ResearchJob({
    required this.id,
    required this.topic,
    required this.depthMinutes,
    required this.status,
    required this.progress,
    this.resultReport,
    required this.createdAt,
    this.updatedAt,
  });

  factory ResearchJob.fromJson(Map<String, dynamic> json) {
    return ResearchJob(
      id: (json['id'] as num).toInt(),
      topic: json['topic'] as String? ?? '',
      depthMinutes: (json['depth_minutes'] as num?)?.toInt() ?? 5,
      status: json['status'] as String? ?? 'queued',
      progress: (json['progress'] as num?)?.toInt() ?? 0,
      resultReport: json['result_report'] as String?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'topic': topic,
        'depth_minutes': depthMinutes,
        'status': status,
        'progress': progress,
        'result_report': resultReport,
        'created_at': createdAt.toIso8601String(),
        'updated_at': updatedAt?.toIso8601String(),
      };

  ResearchJob copyWith({
    String? status,
    int? progress,
    String? resultReport,
  }) =>
      ResearchJob(
        id: id,
        topic: topic,
        depthMinutes: depthMinutes,
        status: status ?? this.status,
        progress: progress ?? this.progress,
        resultReport: resultReport ?? this.resultReport,
        createdAt: createdAt,
        updatedAt: updatedAt,
      );
}
